### PR TITLE
[WIP] EVM trace generation from rust

### LINKED ISSUE #77: Easy to use EVM simulator for unit tests
Currently the way to write unit tests is to manually create the execution steps and resulting operations for all the different op codes each time. This is very tedious. It would be great if we could have a very simple EVM simulator (which would also be updated when implementing an opcode) where we could just write something like this (with maybe some extra ways to mock data for testing purposes):

```
evm.push(1);
evm.push(2);
evm.add();

try_test_circuit!(
   evm.get_execution_steps(),
   evm.get_operations(),
    Ok(())
);
```

I think this will make writing unit tests much more pleasant and faster.

The closest thing to this may be something @CPerezz is working on. Do you think there could be some overlap here @CPerezz? If not I'd be willing to do this myself.

### PR DESCRIPTION
**Pull request by Brechtpd**
Will resolve #77

Note: Currently there are still a lot of differences between the bus-mapping and evm-circuits, and the busmapping witness generation doesn't support many op codes as of now. To make just PUSH1 and STOP work I had to do quite a bit of **temporary hacks**, so please ignore that code. It should be possible to remove these hacks once the circuits use the `ExecutionStep`s and `Operation`s from the busmapping.

Allows easy test writing against geth traces directly from rust:
```
  try_test_circuit!(
      bytecode![
          (PUSH1, BigUint::from(0x10u64)),
          (PUSH2, BigUint::from(0x3040u64)),
          ADD,
          POP
      ],
      Ok(())
  );
```

This is also nice because this also makes sure that the EVM spec is actually correctly implemented (as long as geth is correct) so there is no risk in interpreting the spec not completely correctly and so not actually testing things correctly (which could definitely be a problem for some of the more complex scenario's).

This is done by having rust bindings to geth in the `geth-utils` library crate (using code from https://github.com/appliedzkp/zkevm-testing-vectors/tree/layout_structure/gethutil, I copied the minimal necessary code over so no external dependencies are needed).

To get the full execution trace required for the circuit: bytecode -> geth-utils tracer (geth) -> `bus_mapping::ExecutionTrace::<F>::from_trace_bytes` -> circuit.

### CODE REVIEW

**Comment by CPerezz** [comment]
After reviewing the code, I like the idea behind. But there are some things to discuss I would say.

- I would not make the target of this utility `zkevm-circuits` ie. will not try to include the logic into it or modify the crate as there's a lot of work still to make a good integration with bus-mapping.
- I would try to move the efforts on making this first a tool that provides the info that is stated on each `ExecutionStep` as: Opcode, depth, stack, memory, pc. But I'll not try to add things that aren't even inside of the bus-mapping yet. 

By doing so, we could use this tool as a generator of traces for all of the tests we need and will need to do in the bus-mapping crate without hardcoding the traces.
Therefore, I'd avoid to connect/provide with any fields requested so that we just serve the needs of bus-mapping and we evolve the tool together with the crate on each improvement/refactor.
This would allow to solve your major concerns:
> To make just PUSH1 and STOP work I had to do quite a bit of temporary hacks, so please ignore that code.

This will allow you to further continue the development of the tooling without needing to wait quite some time. 

Also, will help us a lot in order to elaborate tests for all of the opcodes left to impl. 
And finally, we could orient the tool more to end-to-end testing and start to target `zkevm-circuits`.
WDYT @Brechtpd @han0110 @ChihChengLiang @miha-stopar @therealyingtong @rahulbs98

**Comment by Brechtpd** [comment]
Will wait a bit on the feedback of the others that will have a much better understanding of the overall picture. But one thing I want to bring up was that on the community call last week I thought we were going ahead with implementing all the necessary things per opcode together (unless I misunderstood Barry). So a PR for the implementation of an opcode would contain:
1. The python spec implementation
2. The evm circuits implementation
3. The bus mapping implementation (witness generation)

So basically something that ideally would allow end-to-end testing like in this PR. Of course if this is not yet possible nothing we can do about that, but that may complicate things a bit to be able to do the full opcode implementation well.

**Comment by han0110** [comment]
It looks really clean and simple to generate computation trace. Some efforts seem to be overlap with `bus-mapping`, and Carlos's idea sounds good to me.

My only concern is the crate `gobuild` seems to require `go` installed iiuc, but I myself am totally okay with it. (My thought on e2e testing is to pre-generate some trace and mpt proof in json and load them when testing, which also requires to run some go code when generating, so this approach seems more smooth)

**Comment by miha-stopar** [comment]
Cool! It seems to me like a very useful utility. Also, I agree with @CPerezz and @han0110 comments. One minor thing from my side - perhaps to change the examples to be tests instead of having the main function redeclared?

**Comment by barryWhiteHat** [comment]
>    The evm circuits implementation

My thinking was we would instead of building the circuit implementation we would build code to export the witness from geth into the bus mapping. 

https://github.com/appliedzkp/zkevm-circuits/blob/main/bus-mapping/src/evm/opcodes/mload.rs

Is an example of that for mload opcode. 

Its kind of like in libsnark when you have one function to generate constraints and another to generate the witness for that set of constraints.

**Comment by CPerezz** [comment]
> > The evm circuits implementation
> 
> My thinking was we would instead of building the circuit implementation we would build code to export the witness from geth into the bus mapping.
> 
> https://github.com/appliedzkp/zkevm-circuits/blob/main/bus-mapping/src/evm/opcodes/mload.rs
> 
> Is an example of that for mload opcode.
> 
> Its kind of like in libsnark when you have one function to generate constraints and another to generate the witness for that set of constraints.

So I agree in the fact that the circuits and end-to-end testing would not be done with this tool. But the tests for each opcode and more elaborated parsing could indeed be done with this. And it's kinda off the development path that made sense to me. 

WDYT?

**Comment by barryWhiteHat** [comment]
For sure this direction makes sense. Its kind of a situation where @Brechtpd is solving problems before they pop up. We have not really made a concrete push on testing but once we do this tool would be very valuable. 

@CPerezz how do you think we should follow this development path ?

**Comment by ChihChengLiang** [comment]
@Brechtpd I think for the short run for the spec, we can focus on illustrating the idea, like this [example]( https://github.com/appliedzkp/zkevm-specs/blob/master/specs/opcodes/0x50_POP.md). So that people can audit the idea first and then audit if the circuit matches the spec. Making the spec Python code executable or able to producing end-to-end tests at this stage seems to cost a lot of effort.

@CPerezz I like your idea to make trace generation a tool too.

**Comment by Brechtpd** [comment]
> @Brechtpd I think for the short run for the spec, we can focus on illustrating the idea, like this [example](https://github.com/appliedzkp/zkevm-specs/blob/master/specs/opcodes/0x50_POP.md). So that people can audit the idea first and then audit if the circuit matches the spec. Making the spec Python code executable or able to producing end-to-end tests at this stage seems to cost a lot of effort.

Oh yes, I just listed the spec step because it was just a step that was decided on for implementing opcodes. I did not mean that we would immediately use in the testing. With end-to-end testing I only meant also using the witness generation of the busmapping for testing the opcodes in the evm circuit (as they are currently not linked to each other). So for the evm circuit tests just being able to do this easily eventually:
1. specify byte code in rust 
2. get the trace from geth for that byte code
3. get the witness from the busmapping for that trace
4. test the opcode in the circuit using that witness, which tests against geth output (so can be sure the spec is followed correctly) and the busmapping witness (making sure these match, and can be implemented and tested at the same time for ease of development). 

> @CPerezz I like your idea to make trace generation a tool too.

Currently discussing things with him in chat. I think by and large we have the same idea how this would be used, though not really sure about that yet.

File: bus-mapping/src/util.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    v: &F,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+    F: FieldExt,
+{
+    serializer
+        .serialize_str(&BigUint::from_bytes_le(&v.to_bytes()).to_str_radix(16))
+}
```
Comment (CPerezz) [inline bus-mapping/src/util.rs:17]: I belive you don't need to go through all this to generate a FieldElement as `FieldExt`.

Using `EvmWord::{from,to}_bytes_{le,be}()` you should be able to avoid needing the `BigUint` dep as well as maybe even the serde method (Not sure on the last statement).

File: bus-mapping/src/exec_trace.rs [hunk 2]
```diff
@@ -5,38 +5,47 @@ pub(crate) mod parsing;
...
-            gas_limit: F::default(),
-            chain_id: F::default(),
-            base_fee: F::default(),
+            coinbase: F::from_u64(0xc014ba5eu64),
+            timestamp: F::from_u64(1633398551u64),
+            number: F::from_u64(123456u64),
+            difficulty: F::from_u64(0x200000u64),
+            gas_limit: F::from_u64(15_000_000u64),
+            chain_id: F::one(),
+            base_fee: F::from_u64(97u64),
```
Comment (CPerezz) [inline bus-mapping/src/exec_trace.rs:48]: Thanks for adding the proper values! :smile:

File: bus-mapping/src/evm/opcodes/mload.rs [hunk 3]
```diff
@@ -97,81 +97,71 @@ impl Opcode for Mload {
...
+            // Start byte code tested
+            [start]
+            MLOAD
+        };
+
+        let block_ctants = BlockConstants::default();
+
+        // Get the execution steps from the external tracer
+        let obtained_steps = &external_tracer::trace(&block_ctants, &code)
+            [code.get_pos("start")..];
```
Comment (CPerezz) [inline bus-mapping/src/evm/opcodes/mload.rs:141]: I would probably add some method inside `ByteCode` like: `code_from_pos()` or similar. So that we abstract over the Indexing op. 

Just a suggestion anyway :)

File: bus-mapping/src/evm/opcodes/mload.rs [hunk 4]
```diff
@@ -97,81 +97,71 @@ impl Opcode for Mload {
...
-                  "0000000000000000000000000000000000000000000000000000000000000080"
-                ]
-              }
-        ]
-        "#;
+        let code = bytecode! {
+            // Setup state
+            PUSH1(0x80u64)
+            PUSH1(0x40u64)
+            MSTORE
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply bus-mapping/src/evm/opcodes/mload.rs]: Good idea!

I made it possible to call functions on `Bytecode` from within the macro so frequently used code can be implemented in functions. It now looks like this:
```
let code = bytecode! {
    .setup_state()
    
    #[start]
    PUSH1(0x01u64)
    PUSH1(0x02u64)
    ADD

    .call(0xFFFFu64, 0xc0416ac1u64, 0x00u64, 0x00u64, 0x00u64, 0x00u64, 0x00u64)
 
    STOP
};
```

File: bus-mapping/src/evm/opcodes/mload.rs [hunk 5]
```diff
@@ -97,81 +97,71 @@ impl Opcode for Mload {
...
+
+            PUSH1(0x40u64)
+            // Start byte code tested
+            [start]
+            MLOAD
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply bus-mapping/src/evm/opcodes/mload.rs:117]: Added

File: bus-mapping/src/evm/opcodes/mload.rs [hunk 6]
```diff
@@ -97,81 +97,71 @@ impl Opcode for Mload {
...
+            GasInfo {
+                gas: {
+                    let temp = $gas;
+                    $gas -= GasCost::$gas_cost.as_usize() as u64;
+                    temp
+                },
+                gas_cost: GasCost::$gas_cost,
+            }
+        }};
+    }
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply bus-mapping/src/evm/opcodes/mload.rs]: Moved this macro out, but left all other ones currently where they are (I don't want to risk increasing the diff too much in this PR because I may need to do quite a bit of rebasing in the near future :).

File: bus-mapping/src/evm/opcodes/mload.rs (hunk 3 above)
Comment (Brechtpd) [reply bus-mapping/src/evm/opcodes/mload.rs:125]: Here all the code needs to be sent to geth because the code we actually want to test also depends on the memory modifications before it so geth needs to execute as well. So I don't think we can do this range on the `Bytecode` level, it needs to be done on the execution steps. And I didn't want to create dependencies between the two objects so I left it like this. :)

**Comment by Brechtpd** [comment]
> * I think it would be nice to have a Git submodule or subtree pointing to https://github.com/appliedzkp/zkevm-testing-vectors so that we have the examples and also we don't need to copy all of the code.

Still have to do this one, have to find a good way to do it. But I would definitely just move all the code over to a subtree here, submodules are a pain to work with.

> * Would also be really nice if we could adapt/rebase to #116 so that we also capture the storage-related stuff.

Do you want me to do this in this PR? Because that code is still in an external forked repo which makes it a bit messy to already bring into this PR.

File: bus-mapping/src/evm/opcodes/mload.rs (hunk 3 above)
Comment (CPerezz) [reply bus-mapping/src/evm/opcodes/mload.rs:125]: Ohhh I see. 
Makes sense now! Apologies for the missundestanding :)

File: bus-mapping/src/evm/opcodes/mload.rs [hunk 7]
```diff
@@ -97,81 +97,55 @@ impl Opcode for Mload {
...
-                ],
-                "memory": [
-                  "0000000000000000000000000000000000000000000000000000000000000000",
-                  "0000000000000000000000000000000000000000000000000000000000000000",
-                  "0000000000000000000000000000000000000000000000000000000000000080"
-                ]
-              }
-        ]
-        "#;
+        let code = bytecode! {
```
Comment (ed255) [inline bus-mapping/src/evm/opcodes/mload.rs:111]: Here's an idea I would like to discuss.  I think that adding the trace generation into the opcode unit tests has 2 negative sides:
1. Running the tests now requires having go installed and building go packages.
2. While inspecting the unit tests, now it's not obvious what's the trace, because it's being generated in runtime.  So debugging / development can becomes a bit more opaque.

Nevertheless I think having an integrated tool to generate the traces is great, because otherwise it's a very manual process and everyone that needs to add an opcode needs to figure out a way to generate traces, which is not straight forward without having a tool.  So here's my idea:
- have a test folder for evm traces vector tests
- have a rust binary that generates all the evm traces vector tests.  The code for this binary can be extended when new opcodes are added in the bus-mapping.
- In the opcodes unit tests, read the traces vector tests to use as input.
So my proposal is to separate the testing from the test vector generation, and join them by having json files (the traces) in a test vector folder that are created with the test vector generation binary, and are consumed by the opcodes unit tests.  This way it's easy to inspect the trace that a unit file is using, and the generation of the test vector traces don't need to happen during unit testing (eliminating the dependency to go for the general case).  In this proposal, if someone adds code to generate new traces, that person would also commit the new test vectors.  We could even have a github action that generates the test vectors just to verify that the code and the test vectors match.

File: bus-mapping/src/bytecode.rs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    }};
+}
+
+#[macro_export]
+#[doc(hidden)]
+macro_rules! bytecode_internal {
+    // Nothing left to do
+    ($code:ident, ) => {};
+    // PUSHX op codes
+    ($code:ident, $x:ident ($v:expr) $($rest:tt)*) => {{
```
[the remark this answers is not in the document]
Comment (Brechtpd) [reply bus-mapping/src/bytecode.rs:138]: The `PUSH1(0x1)` vs `PUSH1 0x1` is because the latter is harder to parse without additional delimiters per opcode while having `0x1` as an `expr`. So for example this is easy with additional `;`s to mark the different opcodes:

```
PUSH 0x1;
PUSH 0x2;
```
but then we need a lot of `;` on each opcode line... Unless you have a solution for that?

The `#[start]` vs `.start` no real reason, can be changed if you want. :)

File: bus-mapping/src/evm/opcodes/mload.rs [hunk 9]
```diff
@@ -97,81 +97,55 @@ impl Opcode for Mload {
...
         // Generate Step1 corresponding to PUSH1 40
         let mut step_1 = ExecutionStep {
             memory: mem_map.clone(),
             stack: Stack(vec![EvmWord::from(0x40u8)]),
             instruction: OpcodeId::MLOAD,
-            gas_info: GasInfo {
-                gas: 79,
-                gas_cost: GasCost::from(3u8),
-            },
+            gas_info: gas_info!(gas, FASTEST),
```
[the remark this answers is not in the document]
Comment (ed255) [reply bus-mapping/src/evm/opcodes/mload.rs:148]: Your reasoning makes sense, thanks for the clarification!

**Comment by ed255** [comment]
LGTM!  @Brechtpd a few hours ago I merged this PR https://github.com/appliedzkp/zkevm-circuits/pull/116 and it has introduced conflicts with your PR, can you take a look and resolve the conflicts?

**Review by CPerezz — APPROVED** [review]
LGTM.

Brilliant work @Brechtpd . Thank you so much for this contribution! :heart:
